_posts/2025-06-11-neural-sparse-sentence-transformers.md

```diff
@@ -0,0 +1,215 @@
+    "ny": 5.772862434387207,
+    "weather": 4.568415641784668,
+    "now": 3.5895495414733887,
+    "?": 3.331254720687866,
+    "what": 2.7698843479156494,
+    "'": 1.5750715732574463,
+    "s": 1.4272583723068237,
+    "in": 0.49892646074295044,
+    "the": 0.1353016495704651,
+}
+```
+
+#### Encode a document
+
+You can also encode documents the same way, optionally limiting the output to top-weighted tokens:
+
+```python
+# First, use the encode method to convert text to torch.sparse_coo tensor
+# Then, use the decode method to convert torch.sparse_coo tensor to (token, weight) pairs and obtain tokens with top-10 weights
+
+doc_tensor = sparse_model.encode_document("Currently New York is rainy.")
+doc_embedding = sparse_model.decode(doc_tensor, top_k=10)
+
+print(dict(doc_embedding))
+```
+
+**Example output:**
+
+```python
+# this format is compatible with OpenSearch index and search
+
+{
+    'weather': 1.0386661291122437,
+    'york': 1.0295677185058594,
+    'ny': 0.9702811241149902,
+    'rain': 0.9549840092658997,
+    'rainy': 0.9479081630706787,
+    'nyc': 0.8448441624641418,
+    'new': 0.687921404838562,
+    'raining': 0.6790128350257874,
+    'current': 0.6762210130691528,
+    'wet': 0.6448971629142761
+}
+```
+### Step 5: Integrate with OpenSearch
+
+Once you obtained the sparse embeddings from Sentence Transformers, you can index and search them directly in OpenSearch. The output format is fully compatible with OpenSearch's neural sparse search.
+
+For detailed instructions, see the [OpenSearch documentation](https://docs.opensearch.org/docs/latest/vector-search/ai-search/neural-sparse-with-raw-vectors/).
+
+You can also explore example scripts for working with the following embedding types:
+
+- [Sparse embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sparse_encoder/applications/semantic_search/semantic_search_opensearch.py)
+- [Dense embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sentence_transformer/applications/semantic-search/semantic_search_nq_opensearch.py)
+
+
+## OpenSearch's contributions to inference-free models in Sentence Transformers
+
+The OpenSearch team is leading innovation in learned sparse retrieval. Our research in inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.
+
+### IDF-aware Penalty: Enhancing Relevance and Efficiency
+
+Traditional sparse retrievers apply uniform penalties to all tokens during training, ignoring the varying importance of different terms. To address this limitation, OpenSearch researchers introduced IDF-aware penalty[^1], which modulates the penalty based on token significance. This approach applies stronger penalties to common tokens while preserving rare, information-rich tokens that carry more distinctive meaning.
+
```

Comment:
Line 142: Either "an" or "the" should precede "IDF-aware penalty". The 2nd sentence reads as repetitive. An/the IDF-aware penalty modulates the penalty? Can we clarify?